### PR TITLE
New data model

### PR DESCRIPTION
**Pull request by MaciejRewera**
This PR introduces new data model for Submissions and Notifications.
There are not only model changes, but the logic in handling request changed too.

Since the PR is quite huge I would suggest **at least two independent reviews** to be done before merging.

NOTE: Merge needs to be synchronised with frontend merge.

### CODE REVIEW

File: app/uk/gov/hmrc/exports/services/NotificationService.scala [hunk 1]
```diff
@@ -18,71 +18,77 @@ package uk.gov.hmrc.exports.services
...
+  notificationRepository: NotificationRepository
 ) {
 
   private val logger = Logger(this.getClass)
+  private val databaseDuplicateKeyErrorCode = 11000
 
-  def save(notification: DeclarationNotification): Future[Result] = {
-    val eori = notification.eori
-    val convId = notification.conversationId
+  def getAllNotificationsForUser(eori: String): Future[Seq[Notification]] =
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/exports/services/NotificationService.scala]: :+1:

File: app/uk/gov/hmrc/exports/services/NotificationService.scala [hunk 2]
```diff
@@ -18,71 +18,77 @@ package uk.gov.hmrc.exports.services
...
+    Future.sequence(notifications.map(save)).map { seq =>
+      if (seq.exists(_.isLeft)) Left("Failed saving notification")
+      else Right((): Unit)
+    }
+  }
 
-  val PositionFunctionCode = "11"
-  val NameCodeGranted = "39"
-  val NameCodeDenied = "41"
+  def save(notification: Notification): Future[Either[String, Unit]] =
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/exports/services/NotificationService.scala]: :+1:

File: app/uk/gov/hmrc/exports/controllers/NotificationController.scala [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  def getSubmissionNotifications(mrn: String): Action[AnyContent] =
+    authorisedAction(bodyParsers.default) { implicit request =>
+      notificationsService
+        .getNotificationsForSubmission(mrn)
+        .map(notifications => Ok(Json.toJson(notifications)))
+    }
+
+  //TODO response should be streamed or paginated depending on the no of notifications.
+  //TODO Return NO CONTENT (204) when there are no notifications
```
[the remark this answers is not in the document]
Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/controllers/NotificationController.scala]: That was already there. It's unfortunate as it seems that I added the comment.

File: app/uk/gov/hmrc/exports/controllers/SubmissionController.scala [hunk 4]
```diff
@@ -53,99 +93,40 @@ class SubmissionController @Inject()(
...
-                Future.successful(ErrorResponse.ErrorInternalServerError.XmlResult)
-            }
-          case None =>
-            logger.error("Body is not xml")
-            Future.successful(ErrorResponse.ErrorInvalidPayload.XmlResult)
-        }
-      case Left(error) =>
-        logger.error(s"Invalid Headers found. Error message: ${error.message}")
-        Future.successful(ErrorResponse.ErrorGenericBadRequest.XmlResult)
+  private def forwardSubmissionRequestToService(eori: String, vhr: SubmissionRequestHeaders, xml: NodeSeq)(
```
[the remark this answers is not in the document]
Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/controllers/SubmissionController.scala]: Good point! Changed that.

File: app/uk/gov/hmrc/exports/controllers/NotificationController.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (PatrykRudnicki) [reply app/uk/gov/hmrc/exports/controllers/NotificationController.scala]: It was added by me, it's not really important. This thing appear after test review, so I moved information from tests to Controller.

If we're happy with returning 200 and empty sequence we can just remove this TODO

**Pull request merged**